### PR TITLE
HPCC-25351 Add Add WsTopology.TpServiceTargetQuery

### PR DESCRIPTION
**Pull request by wangkx**
The ESP method is added to report the information about services
in a cloud environment.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@GordonSmith ECLWatch UI may use the changes in this PR.

File: esp/scm/ws_topology.ecm [hunk 1]
```diff
@@ -59,7 +59,17 @@ ESPStruct TpCluster
...
+
 //  ===========================================================================
+ESPStruct TpServiceTarget
+{
+    string Name;
+    string Type;
+    unsigned Port;
+    bool TLSSecure;
+    bool Public;
+    string Target;
```
Comment (jakesmith) [inline esp/scm/ws_topology.ecm:71]: what is target? not in any of the service type schema's/or in values.yaml afaics.

File: esp/scm/ws_topology.ecm [hunk 2]
```diff
@@ -59,7 +59,17 @@ ESPStruct TpCluster
...
+};
```
Comment (jakesmith) [inline esp/scm/ws_topology.ecm:72]: How is this info. going to be used?
In particular - if not public, then it is an internal service (used only internally).

I think need to 1st be sure what we want and how this/these services are going to be used.
If it's primarily to display on an eclwatch page for example, apart from *perhaps* ops. looking at the system, I don't think it's useful to display internal services.


Not much detail in the JIRA as to how this info. would be used. Can you expand ?

Out of interest, how does bare-metal expose the esp services running on other ports ?

File: esp/scm/ws_topology.ecm (hunk 1 above)
Comment (wangkx) [reply esp/scm/ws_topology.ecm:71]: I saw the target from the roxie service:
services:
      - name: roxie
        type: roxie
        port: 9876
        target: roxie

File: esp/scm/ws_topology.ecm (hunk 2 above)
Comment (wangkx) [reply esp/scm/ws_topology.ecm:72]: A user may use the port: 8010, 8002, etc. Should a user need to know the TLS?
If we only want to expose public services, I can add the filter. Should we let Admin users know internal services?
In bare-metal mode, we display similar information on Operation/System Servers page.

Comment (jakesmith) [reply esp/scm/ws_topology.ecm:72]: I suspect we don't want to show internal services by default, not sure in general.
If it's useful, can be acted upon, then it makes sense I think, if not, it either needs to be ignored, or shown separately somehow.
I'd just show the public services for now.

File: esp/scm/ws_topology.ecm (hunk 1 above)
Comment (jakesmith) [reply esp/scm/ws_topology.ecm:71]: right ok, I see. The target is the roxie, the service name of this roxie is also named roxie.

**Comment by wangkx** [comment]
@jakesmith I have changed the code to only return public services. I also removed the target attribute since it only exists in roxie service and the roxie service is not a public service.

**Comment by jakesmith** [comment]
>I have changed the code to only return public services. I also removed the target attribute since it only exists in roxie service and the roxie service is not a public service.

Actually, the roxie service is public.. It's the public (client) gateway to the roxie that's deployed.
NB: see kubectl get svc 

But that's effectively a bug in how services are generated, i.e. should generate public for this service too.
Please open a JIRA for that - and link it to this one.

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 3]
```diff
@@ -1970,6 +1970,36 @@ void CTpWrapper::getAttPath(const char* Path,StringBuffer& returnStr)
...
+
+        //Only show the public services for now
+        if (!service.getPropBool("@public"))
+            continue;
+
+        Owned<IEspTpServiceTarget> serviceTarget = createTpServiceTarget();
+        serviceTarget->setName(name);
+        serviceTarget->setType(type);
+        serviceTarget->setPort(service.getPropInt("@port"));
+        if (service.hasProp("@tls")) //if not set, default to ?
```
Comment (jakesmith) [inline esp/smc/SMCLib/TpWrapper.cpp:1995]: I would remove the comment, I think it's fine to treat absence of this property as false.
So would prob. change to:
```
   if (service.getPropBool("@tls"))
```

**Review by jakesmith — COMMENTED** [review]
@wangkx - looks fine, 1 minor comment

File: esp/smc/SMCLib/TpWrapper.cpp (hunk 3 above)
Comment (wangkx) [reply esp/smc/SMCLib/TpWrapper.cpp:1995]: Will change

**Comment by wangkx** [comment]
> But that's effectively a bug in how services are generated, i.e. should generate public for this service too. Please open a JIRA for that - and link it to this one.

https://track.hpccsystems.com/browse/HPCC-25382

**Review by jakesmith — COMMENTED** [review]
@wangkx - looks good, please squash

**Comment by wangkx** [comment]
@jakesmith I did rebase and tests.

**Comment by jakesmith** [comment]
@wangkx - looks good, thought of one more question though.

is getServiceTargets correct naming?

These are services, I think we're calling the main component queues "Targets".
So it may be better for these methods to be just called "getServices" without conflating with Targets.

Not sure. @ghalliday - what are your thoughts?

**Comment by wangkx** [comment]
The existing WsTopology service already has a method named TpServiceQuery which reports the hpcc components (dali, esps, dfuservers, etc) in bare metal env. Should I merge the code there?

**Comment by jakesmith** [comment]
> The existing WsTopology service already has a method named TpServiceQuery which reports the hpcc components (dali, esps, dfuservers, etc) in bare metal env. Should I merge the code there?

What is that used for/by ? It may not be very applicable to cloud, especially if esp services are now being returned from this new service.

**Comment by wangkx** [comment]
> What is that used for/by ? It may not be very applicable to cloud, especially if esp services are now being returned from this new service.
For example, http://10.173.160.101:8010/#/stub/OPS-DL/SystemServersQuery-DL/SystemServersQueryWidgetIframeWidget
Mainly used by Operation team. The page shows Network Address, working folders, links to log files and configuration file. Gordon may hide it for clouds.

**Comment by jakesmith** [comment]
> For example, http://10.173.160.101:8010/#/stub/OPS-DL/SystemServersQuery-DL/SystemServersQueryWidgetIframeWidget
> Mainly used by Operation team. The page shows Network Address, working folders, links to log files and configuration file. Gordon may hide it for clouds.

ok, then the 'services' in a cloud setup (i.e. those returned by this service and generated in the component config) are different.
And yes TpServiceQuery sounds like it should be suppressed in cloud.
I'm not sure it's just eclwatch though - what sense will this info. make in cloud world, if called by anything else?

If this (and other) services replace TpServiceQuery (and others), then they can be excluded from each build.

I suspect there's quite a few other things in WsTopology that don't quite make sense in cloud anymore, e.g. TpMachine/tpMachineType, or things with paths to logs, etc.
At some point it might make more sense to have a WsContainerServices for services calls like this one, leaving the bare-metal ones as they are.

**Comment by wangkx** [comment]
> At some point it might make more sense to have a WsContainerServices for services calls like this one, leaving the bare-metal ones as they are.

Considering a possible future WsContainerServices (or a better name) service, it may be better not merge the code into the existing WsTopology.TpServiceQuery. We may give the TpServiceTargetQuery method a better name (TpContainerServiceQuery?) and, later on, move the whole method, and other methods (like the DropZone related methods) into the WsContainerServices service.

**Comment by jakesmith** [comment]
> > At some point it might make more sense to have a WsContainerServices for services calls like this one, leaving the bare-metal ones as they are.
> 
> Considering a possible future WsContainerServices (or a better name) service, it may be better not merge the code into the existing WsTopology.TpServiceQuery. We may give the TpServiceTargetQuery method a better name (TpContainerServiceQuery?) and, later on, move the whole method, and other methods (like the DropZone related methods) into the WsContainerServices service.

It depends how much the existing WsTopology services and new container services overlap/diverge.
If it a lot of of WsTopology still makes sense, then it may make sense to be conjoined, trying to minimize the differences.
If they're quite different in many areas, and the containerized services are fewer, smaller and cleaner in a separate service, I suspect it is better to separate.

**Comment by jakesmith** [comment]
@GordonSmith - can you take look at this from a functionality point of view.
This service calls currently returns a list of services that are public, i.e. open to the outside (there are other internal services that it does not show). NB: these are the 'services' generated into the global config.

What will eclwatch + other client want to see etc. ?
Is the info. per service adequate?

I think there's another JIRA re. a separate service call that will return the job queues that clients can submit to.

**Review by jakesmith — COMMENTED** [review]
@wangkx - should this be closed / does https://github.com/hpcc-systems/HPCC-Platform/pull/14636 replace it?

**Comment by wangkx** [comment]
Looks like this PR should be replaced by https://github.com/hpcc-systems/HPCC-Platform/pull/14636.

**Pull request closed without merging**